src/transformers/models/detr/modeling_detr.py

```diff
@@ -0,0 +1,2277 @@
+        )
+
+    def forward(self, x):
+        # move reshapes to the beginning
+        # to make it user-friendly
+        weight = self.weight.reshape(1, -1, 1, 1)
+        bias = self.bias.reshape(1, -1, 1, 1)
+        running_var = self.running_var.reshape(1, -1, 1, 1)
+        running_mean = self.running_mean.reshape(1, -1, 1, 1)
+        epsilon = 1e-5
+        scale = weight * (running_var + epsilon).rsqrt()
+        bias = bias - running_mean * scale
+        return x * scale + bias
+
+
+def replace_batch_norm(m, name=""):
+    for attr_str in dir(m):
+        target_attr = getattr(m, attr_str)
+        if isinstance(target_attr, torch.nn.BatchNorm2d):
+            frozen = FrozenBatchNorm2d(target_attr.num_features)
+            bn = getattr(m, attr_str)
+            frozen.weight.data.copy_(bn.weight)
+            frozen.bias.data.copy_(bn.bias)
+            frozen.running_mean.data.copy_(bn.running_mean)
+            frozen.running_var.data.copy_(bn.running_var)
+            setattr(m, attr_str, frozen)
+    for n, ch in m.named_children():
+        replace_batch_norm(ch, n)
+
+
+class TimmBackbone(nn.Module):
+    """
+    Timm convolutional backbone.
+
+    nn.BatchNorm2d layers are replaced by FrozenBatchNorm2d as defined above.
+
+    """
+
+    def __init__(self, name: str, dilation: bool, return_intermediate_layers: bool):
+        super().__init__()
+
+        kwargs = {}
+        if dilation:
+            kwargs["output_stride"] = 16
+        backbone = create_model(name, pretrained=True, features_only=True, out_indices=(1, 2, 3, 4), **kwargs)
+        # replace batch norm by frozen batch norm
+        with torch.no_grad():
+            replace_batch_norm(backbone)
+        self.body = backbone
+        self.return_intermediate_layers = return_intermediate_layers
+        self.intermediate_channel_sizes = self.body.feature_info.channels()
+
+        if "resnet" in name:
+            for name, parameter in self.body.named_parameters():
+                if "layer2" not in name and "layer3" not in name and "layer4" not in name:
+                    parameter.requires_grad_(False)
+
+    def forward(self, pixel_values: torch.Tensor, pixel_mask: torch.Tensor):
+        # send pixel_values through the body to get list of feature maps
+        feature_maps = self.body(pixel_values)
+
+        if not self.return_intermediate_layers:
+            # get final feature map
+            feature_map = feature_maps[-1]
+            # downsample the pixel_mask to match the shape of the feature map
+            mask = F.interpolate(pixel_mask[None].float(), size=feature_map.shape[-2:]).to(torch.bool)[0]
+            return feature_map, mask
+        else:
+            out = []
+            for i, x in enumerate(feature_maps):
+                mask = F.interpolate(pixel_mask[None].float(), size=x.shape[-2:]).to(torch.bool)[0]
+                out.append((x, mask))
+            return out
+
+
+class Joiner(nn.Sequential):
```

Comment:
`DETRJoiner` ? Also a bit weird the name IMO

Reply:
Not a big fan of this class. 

1. Inheriting from `nn.Sequential` and using `self[0]` makes it super hard to understand & debug
2. If the whole forward method is wrapped into a `if ... else ...` statement I think we should have two classes instead with the proper names and have the `if else statement` further below

Reply:
Yeah the Joiner was taken from the original implementation, and your comment makes total sense! Will make it more simple and easier to read.